### ISSUE TITLE
zero- and one-element tuples

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#883]
@gafter commented on [Fri Jan 06 2017](https://github.com/dotnet/roslyn/issues/16313)

To best align positional pattern-matching (which won't restrict the number of values being deconstructed) with tuples, please support zero-element and one-element tuple types and tuple expressions.

---

@AdamSpeight2008 commented on [Fri Jan 06 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-271036820)

Doesn't the current tuple parser require a minimum of 2 values?

---

@alrz commented on [Fri Jan 06 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-271049625)

@AdamSpeight2008 Yes.

---

@bbarry commented on [Fri Jan 06 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-271049862)

Why not align the other way, don't recognize deconstruction syntax with fewer than 2 elements?

---

@HaloFour commented on [Fri Jan 06 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-271059298)

@gafter 

I assume that this will be a post-C# 7.0 consideration?

@bbarry 

Without support for conditional deconstruction of zero/one values then patterns like `None()` and `Some(T)` would be impossible.

---

@orthoxerox commented on [Sat Jan 07 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-271073828)

@bbarry Single-value tuples are necessary for primitive wrapper types.

---

@dsaf commented on [Mon Jan 09 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-271316948)

Would leveraging it for naming primitive return values be considered an abuse?

```csharp
(int id) SaveSomethingToDb(Something smth);
```

instead of

```csharp
int SaveSomethingToDbAndReturnId(Something smth);
```

---

@HaloFour commented on [Mon Jan 09 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-271412441)

I like how Swift handles "womples", they're effectively just the scalar value itself.  So an `Int32` is an `(Int32)` is an `((Int32))`.  This makes sense as `1`, `(1)` and `((1))` are all the same value.  Swift does not allow for naming the element of a "womple", though, which kind of makes sense.

---

@Richiban commented on [Tue Jan 31 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-276357969)

Will the zero-tuple be explicitly referred to as unit anywhere?

I'd love to see some support for this in the future, such as being able to use `Func<int, ()>` instead of `Action<int>`, for example.

---

@alrz commented on [Tue Jan 31 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-276360288)

I think the type (and literal) for unit could be `()` if it wasn't ambigious, but using it in place of void probably requires language support so you dont need to explicitly return `()`.

---

@jcouv commented on [Tue Sep 05 2017](https://github.com/dotnet/roslyn/issues/16313#issuecomment-327332707)

@gafter Is this tracking a compiler/API change or a language feature? If the latter, could you close or move to csharplang?

**Comment by yaakov-h** [comment]
What's the purpose of a zero-element tuple?

**Comment by mattwar** [comment]
Purpose.. probably to satisfy a desire for a unit type/value that some functional languages have (as opposed to 'void' that C#/.net has), and some language's common convention of using empty collections or zero-valued tuples to represent nothing, aka unit.

**Comment by HaloFour** [comment]
There might also be a need to support them conceptually to permit deconstruction of zero and one element patterns, e.g. `Some(T)` and `None()`.

**Comment by DavidArno** [comment]
@yaakov-h,

As @mattwar says, `()` is another way of saying `unit`, ie no value.

However, I feel it would be a bad idea to introduce yet another type (`unit`) to C# to represent that (no) value, as we already have two other ways of indicating no value: `null` and `void`. I'd far prefer to see the language support the idea of void expressions (eg, see #135) and to have the syntax, `()`, be a way of expressing `void` in such expressions:
```cs
void Foo(bool b) => b ? Console.WriteLine("True!") : ();
```
Having said that. I'm not really convinced that `()` adds anything that couldn't be covered by just using `void`:
```cs
void Foo(bool b) => b ? Console.WriteLine("True!") : void;
```
 but I may be missing some use-case for it.

**Comment by AdamSpeight2008** [comment]
I'm thinking that they are, wanting to be able to treat `Tuple` and `ValueTuple` as a pseudo-base class.
```c#
tuple result = SomeFunction();
switch( result )
{
  case Is (): /* roughly MayBe.None */
  case Is (var A): /* roughly MayBase.Some<T> */
  case Is (var A, var B):
}
```
with `SomeFunction` *roughly* being
```c#
tuple SomeFunction()
{
  If( condition )
  {
    return ( Value );
  }
  else
  {
   return ();
  }
}
```

**Comment by scalablecory** [comment]
If you have:

    void Foo((int Bar) x);

How would you call it? Would it implicitly convert `Foo(1)`? You can't use any number of parens like `Foo((1))` to disambiguate it.

**Comment by HaloFour** [comment]
@scalablecory 

I like how Apple Swift does it.  A single-element tuple (womple) is just an alternative syntax for a scalar value.  An `(Int)` is just an `Int`, which kind of makes sense syntactically as `(123)` is really just `123`.  But one of the side-effects is that unlike tuples of two or more elements, you can't name that single element.

**Comment by scalablecory** [comment]
@HaloFour even if it does boil down to `int` I think feature parity (naming deconstruction etc.) could all be accomplished with the help of attributes.

**Comment by jnm2** [comment]
> But one of the side-effects is that unlike tuples of two or more elements, you can't name that single element.

I consider this a feature though. 😆

**Comment by quinmars** [comment]
> but I may be missing some use-case for it.

A real unit type would be very useful for generic types. Take a look on tasks. We have there two types in C#, one is `Task` the other `Task<T>`. With a language featured unit type from the beginning, `Task` would have been simply `Task<()>` or `Task<void>`.

For tasks this is already to late. But think of async enumerables. If  `IAsyncEnumerable<T>` is the sequential counterpart of `Task<T>`, what is the counterpart for `Task`? I doubt there will be an non-generic `IAsyncEnumerable`, because it's not worth the effort. On the other hand we have to life then with the lack of symmetry.

**Comment by DavidArno** [comment]
@quinmars,

I agree that a real unit type would be useful. If it's feasible though, then I'd like to see `void` become that unit type, so that I could do `Task<void>` or ` IAsyncEnumerable<void>`.

Of course, it may not be practicable. In that case, I think the next best solution is a `Unit` type (and `unit` keyword alias), but that has similar semantics to `void`, especially with regards to `return`:
```cs
unit Foo() 
{
    Console.WriteLine("foo");
} // no return needed as unit has no value to return
```

**Comment by YairHalberstadt** [comment]
@DavidArno 
I think void should be the alias for the unit type ().

**Comment by jnm2** [comment]
>  I doubt there will be a non-generic IAsyncEnumerable, because it's not worth the effort.

But let's assume we're willing to put any amount of effort into it. What would it do, exactly? Lack a `Current` property?

**Comment by Joe4evr** [comment]
> In that case, I think the next best solution is a `Unit` type (and `unit` keyword alias), but that has similar semantics to `void`, especially with regards to `return`:

So basically, exactly what I described [here](https://github.com/dotnet/csharplang/issues/696#issuecomment-310541444).

**Comment by HaloFour** [comment]
@YairHalberstadt 

> I think void should be the alias for the unit type ().

`void` already means something quite different in C#, and is valid syntax in some scenarios where it would collide with any unit type, e.g. `typeof(void)`.

**Comment by quinmars** [comment]
> > In that case, I think the next best solution is a Unit type (and unit keyword alias), but that has similar semantics to void, especially with regards to return:
> 
> So basically, exactly what I described here.

It appears to be the best what we can get. And I like it. The following feels very natrual:

```csharp
async IAsyncEnumerable<unit> ToAsyncEnumerable(IEnumerable<Task> list)
{
    foreach (var t in list)
    {
        await t;
        yield return;
    }
} 
```

and:

```csharp
// T can be unit!
async IAsyncEnumerable<T> ToAsyncEnumerable<T>(IEnumerable<Task<T>> list)
{
    foreach (var t in list)
    {
        yield return await t;
    }
} 
```

>  >  I doubt there will be a non-generic IAsyncEnumerable, because it's not worth the effort.

>But let's assume we're willing to put any amount of effort into it. What would it do, exactly? Lack a `Current` property?

Probably.

**Comment by yaakov-h** [comment]
Non-generic `IEnumerable` is an enumerable of result `object`s. Making `IAsyncEnumerable` an enumerable of non-results seems strange and asymmetric.

**Comment by jnm2** [comment]
If there is no result being produced for each MoveNext, why would anyone want to call `MoveNextAsync` in a loop until it returns false versus just awaiting a non-generic `Task`?

**Comment by quinmars** [comment]
@jm2, an `IAsyncEnumeration<unit>` is not a just a collection of the same useless value, it has a second dimension: the timing.

Let's say you want to write an email to a list of recipients via `Task SendEmailAsync(string r)`, but only for 5 seconds. You could write:

```csharp
var lastOrder = DateTime.Now + TimeSpan.FromSeconds(5);
var orders = recipients
    .Select(r => SendEmailAsync(r)) // due to lazy evaluation the mails are not sended now
    .ToAsyncEnumerable() // The (extension) method in my last post
    .TakeWhile(_ => DateTime.Now < lastOrder);

foreach async (var _ in orders)
{
    // We can write this line after every email sended
    // not just when all emails are sended.
    Console.WriteLine($"Email sended at {DateTime.Now}");
}
```

**Comment by quinmars** [comment]
@jm2 and @yaakov-h, don't get me wrong.  I do not think that it is reasonable to add a non-generic `IAsyncEnumerable`. I just wanted to show that there is a symmetry gap, which an unit type could fill.

**Comment by Neme12** [comment]
I don't think one-element tuples should be in any way "equivalent" to the value itself. If you have any other tuple, it's just a container type and to get the actual value, you have to dot into it. For any other tuple, you have to dot into it and can even give it a custom name. Why should it be different for a one-tuple?

**Comment by HaloFour** [comment]
@Neme12 

Tuples are only somewhat containers in the C# language.  You have to dot into them out of necessity, but in some cases the language does (or will) treat them just as separate values and ignore that container.  For example, the compiler already completely skips the container if you immediately deconstruct a tuple literal.  This is despite the fact that the `ValueTuple<...>` implementation could technically do something custom in its constructor.  There is also a championed proposal to have the compiler allow equality comparisons between tuples by emitting direct comparisons between the tuple elements, again skipping any potential equality operator implementation on `ValueTuple<...>`:

```cs
(int, int) GetPoint(int x, int y) => (x, y);
(float, float) GetPointF(float x, float y) => (x, y);

var x = GetPoint(1, 2);
var y = GetPointF(1.0f, 2.0f);

if (x == y) { ... }
// converted to
if (x.Item1 == y.Item1 && x.Item2 == y.Item2) { ... }
```

IIRC, in math a "womple" is just the single value.  In various programming languages (e.g. Swift) that's also the case.  There are probably advantages and disadvantages to both approaches.  I'm not particularly vested in either, but I'd like to hear scenarios where womples are actually useful/necessary.

**Comment by Neme12** [comment]
@HaloFour My point is you still have to deconstruct it to get the value.
I don't see why the pattern shouldn't continue:
```c#
var a = (item1: 1, item2: 2, item3: 3,).item1;  // OK
var b = (item1: 1, item2: 2).item1;    // OK
var c = (item1: 1).item1;    // error?

```
That would really be inconsistent for the one-tuple to behave completely differently.

I can't think of any usage for a 1-tuple, I'm more interested in 0-tuples, and it would definitely be weird to have 0 and not 1. I think they should be allowed even if it's just for consistency.

> IIRC, in math a "womple" is just the single value. 

C# is not math (and I'm quite happy about that) 😄

**Comment by mcintyre321** [comment]
I'd like `var c = (item1: 1);    // type == ValueTuple<int>` to be available.

I've found a way of serializing the tuple property names, and it's quite annoying not to be able to use "single value + name" tuples.

**Comment by ufcpp** [comment]
```cs
Func<(T1 x1, T2 x2)>
Func<(T1 x1)>
Func<()> // I really want this
...

Task<(T1 x1, T2 x2)>
Task<(T1 x1)>
Task<()> // I really want this
...
```

**Comment by mcintyre321** [comment]
Does anyone know if this will be considered at any point? I keep having to use a dummy parameter in certain situations e.g. class SomeCommand : Command<(string someArgument, object _)> { ... } ...

**Comment by jcouv** [comment]
@mcintyre321 Can you expand on your scenario? Why not just do `class SomeCommand : Command<string> ...`?

**Comment by mcintyre321** [comment]
I want the value to be named, not just for sugar, but for serialisation as well (I have some code that will read the name attributes).

**Comment by Joe4evr** [comment]
Sounds like a bad case of Round Peg, Square Hole. If you want named values, *especially* for the purpose of serialization, it's much more appropriate to use a `Dictionary<string, Command>`.

**Comment by mcintyre321** [comment]
I probably haven't explained my use cases sufficiently. I find there are LOTS of situations where I would like to create a class where one would use a (potentially deeply nested) ValueTuple as a generic parameter, e.g. in a `MessageHandler<TIn, TOut>` which works fine, except when I have single valued items. 

or `public class EngineTemperature : ValueOf<int> { }` vs `public class EngineTemperature : ValueOf<(int Celsius)> { }`. I find the latter more helpful, but it won't compile. (NB I'm using my ValueOf library here, but this will be a problem with Record Types once they are part of the language).

Maybe these use cases aren't that common (yet - I think the Record Type use case is very desirable as you will be able to define structures for API messages very concisely), but the fact that single item ValueTuples have different rule (e.g. you have to use the class explicitly (as in `public ValueTuple<int> GetSomething(){...}`, rather than being able to do `public (int) GetSomething(){...)` complicates the language, and increases code churn when refactoring.

**Comment by mcintyre321** [comment]
edit: thinking about it, I don't think it can be done without clashing with redundant parenthesis in existing code (for single valued, unnamed tuples)

```
public void Foo (Bar bar) { ... }
...
Foo((someBar));
```

The compiler could probably spot a single value named tuple though.

**Comment by bandleader** [comment]
@mcintyre321 I think that's one the main reasons they haven't added 1-tuples. 
`(5)` is currently valid C# and resolves to `5`.

It's difficult to remove the ambiguity elegantly:
```c#
(item1: 5) // By naming the element?
(: 5) // Similar but skipping the name?
(5) // This would still return 5
```

An alternative would be to disambiguate contextually, after the compiler has finished binding and knows the types. That is, if you pass a parenthesized expression of type `T` to a place where `(T1)` is expected, lower the syntax to a tuple. It's not _really_ a breaking change unless an existing method takes a `ValueTuple<T1>` (which does exist) with another overload that takes a plain `T1` (which is a strange scenario to want to disambiguate). **However,** having code that _parses_ differently based on types is usually a bad idea.

**Comment by lucasteles** [comment]
Will we suport deconstruct inline single value tuples?

Today in C#8 beta we suport it only in pattern matching

For exemple
```cs
        public abstract class Option<T> { }
        public class None<T> : Option<T> { }
        public class Some<T> : Option<T>
        {
            public T Value { get; }
            public Some(T value) => Value = value;

            public void Deconstruct(out T value) => value = Value;
        }
```

This works:
```cs
            if (maybeFoo is Some<Foo>(var value))
            {  
                 // value is a value of type Foo
                ...
            }
```

but this wont
```cs
            if (maybeFoo is Some<Foo> someFoo)
            { 
               var (value) = someFoo;
                ...
            }
```

I think this is something thats should work

**Comment by zacharylayne** [comment]
I'd like to see this too. I'm putting together a scripting framework that wraps .NET's scripting library, and I'm considering letting scripts declare tuples for expected input (e.g. `(string foo, char bar) input;`. If 1-tuple syntax doesn't work, this is a less attractive option.

**Comment by myblindy** [comment]
This would be nice in general, it could allow for code such as this:

```C#
Dictionary<string, (int color)> dict;

// later
var color = dict["Table"].color;  
```

Compared to just `dict["Table"]` which doesn't provide a hint at a glance as to what it returns, even with a better dictionary name. Maybe I basically just want named dictionary value types, but this is by far easier to implement.

**Comment by CyrusNajmabadi** [comment]
I think that code isn't clear because you called it `dict`. Just call it 'nameToColor' and it will be clear.

**Comment by jnm2** [comment]
I like `colorsByName` because then each entry is `colorByName`. `colorByName.Key` is obviously 'name' and `colorByName.Value` is obviously 'color.'

**Comment by Cephei** [comment]
This was very confusing when I attempted use a Deconstruct() with a single out param, `n`, but the compiler would not allow me to pattern match using this. The error just says `An expression of type 'xyz' cannot be handled by a pattern of type 'n'.`

**Comment by Shadowblitz16** [comment]
single element tuples should be able to be named as it makes good quick objects in generics

Also regarding tuple syntax while wrapping a value inside braces is valid it doesn't actually do anything unless it is a logical expression.

Just check whether the whole expression is wrapped if it is then its a tuple if its only partially wrapped it's not.

there is no reason to do something like...
```cs
var value = (x * y + z == 0 ? 0 : 1)
```

However it is still useful to do things like..
var value  = (x * y + z) == 0) ? 0 : 1
var tuple = ((x * y + z) == 0) ? 0 : 1)
var value = x * (y + z)
var tuple  = (x * y + z)

**Comment by CyrusNajmabadi** [comment]
@Shadowblitz16 That code already has meaning and exists today in many projects.  We cannot break it by having its meaning change.

**Comment by binki** [comment]
I wonder if target typing could be used when passing a single-value tuple via a method call or doing an assignment. That approach was used recently to make the ternary operator more usable.

It would be unfortunate if I had to write this code:

```csharp
(int Item1) GetInfo() {
  return (Item1: 2);
}
```

When the compiler could infer what I mean with this code:

```csharp
(int Item1) GetInfo() {
  return (2);
}
```

But, even if I do have to write `(Item1: 2)`, I still want this feature. I oftentimes am authoring a method and changing it between returning a single value and multiple values. Not being able to keep it a tuple for the one-value case makes that take a lot more work. Also, sometimes I want a name for clarity and end up making a garbage second member of the tuple just to get that:

```csharp
(int Count, int _unused) GetInfo() {
  return (2, 0);
}
```

Using a `struct record` is a good alternative (now that it exists, yay!), but even that is a little bit heavyweight for some use cases.

I also admit that I think that Python’s way of solving this problem with a trailing comma is acceptable. I do not expect this to be acceptable to the C# community because it looks wrong:

```
>>> 1,
(1,)
>>> type((1,))
<class 'tuple'>
```

**Comment by jez9999** [comment]
Another use-case for supporting shorthand single-element tuples:

```
public void DoThing(int id, Func<(int Id1, int Id2, string Message), bool> fnProcessCallback) {
    //...
    fnProcessCallback(1, 2, "hw");
}

//...

foo.DoThing(12345, (args) => {
    var (id1, id2, msg) = args;
}
```

Effectively allows names arguments for callbacks, as the arguments are all just one tuple with names.  Works fine until you need a callback with just one argument:

```
Func<(int Id1), bool>
//...
var (id1) = args;
```

Would be nice for this to work.
